File: go/internal/upstream/pool.go
package upstream

import (
	"context"
	"net"
	"sync/atomic"
)

// Pool is a bounded set of reusable TCP connections to a single upstream
// address (phase 2: stop dialing per request). Connections are dialed
// lazily on demand, up to maxSize concurrently outstanding; eviction is
// error-driven — a connection that errors during use is closed and not
// returned to the pool, so the next Get dials fresh.
type Pool struct {
	addr        string
	conns       chan net.Conn
	maxSize     int
	outstanding atomic.Int32
}

// NewPool returns a Pool for addr that allows up to maxSize concurrently
// outstanding connections.
func NewPool(addr string, maxSize int) *Pool {
	return &Pool{
		addr:    addr,
		conns:   make(chan net.Conn, maxSize),
		maxSize: maxSize,
	}
}

// Get returns an idle pooled connection if one is available, dials a new
// one if the pool is under maxSize, or blocks until a connection is
// returned (or ctx is done) if the pool is at capacity.
func (p *Pool) Get(ctx context.Context) (net.Conn, error) {
	select {
	case conn := <-p.conns:
		return conn, nil
	default:
	}

	if p.outstanding.Add(1) <= int32(p.maxSize) {
		conn, err := Dial(p.addr)
		if err != nil {
			p.outstanding.Add(-1)
			return nil, err
		}
		return conn, nil
	}
	p.outstanding.Add(-1)

	select {
	case conn := <-p.conns:
		return conn, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

// Put returns conn to the pool if healthy is true, or closes it and frees
// its slot otherwise.
func (p *Pool) Put(conn net.Conn, healthy bool) {
	if !healthy {
		conn.Close()
		p.outstanding.Add(-1)
		return
	}

	select {
	case p.conns <- conn:
	default:
		// Pool full (shouldn't happen if accounting is correct) — drop it.
		conn.Close()
		p.outstanding.Add(-1)
	}
}

// Manager holds one Pool per upstream address.
type Manager struct {
	pools map[string]*Pool
}

// NewManager returns a Manager with one Pool per addr in addrs, each
// allowing up to sizePerUpstream outstanding connections.
func NewManager(addrs []string, sizePerUpstream int) *Manager {
	pools := make(map[string]*Pool, len(addrs))
	for _, addr := range addrs {
		pools[addr] = NewPool(addr, sizePerUpstream)
	}
	return &Manager{pools: pools}
}

// Pool returns the Pool for addr.
func (m *Manager) Pool(addr string) *Pool {
	return m.pools[addr]
}

// ForwardPooled forwards req using a connection from pool, returning the
// connection to the pool on success or discarding it on error so the next
// Get dials fresh.
func ForwardPooled(ctx context.Context, pool *Pool, req *Request) (*Response, error) {
	conn, err := pool.Get(ctx)
	if err != nil {
		return nil, err
	}

	resp, err := Forward(conn, req)
	if err != nil {
		pool.Put(conn, false)
		return nil, err
	}

	pool.Put(conn, true)
	return resp, nil
}

// Forwarder forwards a request to addr, abstracting over per-request
// dialing (phases 0/1a/1b) vs. pooled connections (phase 2+).
type Forwarder interface {
	Forward(addr string, req *Request) (*Response, error)
}

// DialForwarder forwards each request over a freshly dialed connection
// (phases 0/1a).
type DialForwarder struct{}

// Forward implements Forwarder.
func (DialForwarder) Forward(addr string, req *Request) (*Response, error) {
	return ForwardHTTP(addr, req)
}

// PoolForwarder forwards each request using Manager's pooled connections
// (phase 2).
type PoolForwarder struct {
	Manager *Manager
}

// Forward implements Forwarder.
func (f PoolForwarder) Forward(addr string, req *Request) (*Response, error) {
	return ForwardPooled(context.Background(), f.Manager.Pool(addr), req)
}

File: crates/service-proxy/src/upstream/pool.rs
//! Bounded per-upstream connection pool (phase 2+).

use std::collections::HashMap;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use tokio::net::TcpStream;
use tokio::sync::{Mutex, Notify};

use super::dial::{dial, forward, Request, Response, UpstreamError};

pub struct Pool {
    addr: String,
    max_size: usize,
    idle: Mutex<Vec<TcpStream>>,
    outstanding: AtomicUsize,
    notify: Notify,
}

impl Pool {
    pub fn new(addr: String, max_size: usize) -> Self {
        Self {
            addr,
            max_size,
            idle: Mutex::new(Vec::new()),
            outstanding: AtomicUsize::new(0),
            notify: Notify::new(),
        }
    }

    pub async fn get(&self) -> Result<TcpStream, UpstreamError> {
        loop {
            {
                let mut idle = self.idle.lock().await;
                if let Some(conn) = idle.pop() {
                    return Ok(conn);
                }
            }

            let prev = self.outstanding.fetch_add(1, Ordering::AcqRel);
            if prev < self.max_size {
                match dial(&self.addr).await {
                    Ok(conn) => return Ok(conn),
                    Err(e) => {
                        self.outstanding.fetch_sub(1, Ordering::AcqRel);
                        return Err(e);
                    }
                }
            }
            self.outstanding.fetch_sub(1, Ordering::AcqRel);

            self.notify.notified().await;
        }
    }

    pub async fn put(&self, conn: TcpStream, healthy: bool) {
        if !healthy {
            drop(conn);
            self.outstanding.fetch_sub(1, Ordering::AcqRel);
            self.notify.notify_one();
            return;
        }

        let mut idle = self.idle.lock().await;
        if idle.len() < self.max_size {
            idle.push(conn);
            drop(idle);
            self.notify.notify_one();
        } else {
            drop(idle);
            drop(conn);
            self.outstanding.fetch_sub(1, Ordering::AcqRel);
            self.notify.notify_one();
        }
    }
}

pub struct Manager {
    pools: HashMap<String, Arc<Pool>>,
}

impl Manager {
    pub fn new(addrs: &[String], size_per_upstream: usize) -> Self {
        let mut pools = HashMap::with_capacity(addrs.len());
        for addr in addrs {
            pools.insert(addr.clone(), Arc::new(Pool::new(addr.clone(), size_per_upstream)));
        }
        Self { pools }
    }

    pub fn pool(&self, addr: &str) -> Option<Arc<Pool>> {
        self.pools.get(addr).cloned()
    }
}

pub async fn forward_pooled(pool: &Pool, req: &Request) -> Result<Response, UpstreamError> {
    let mut conn = pool.get().await?;
    match forward(&mut conn, req).await {
        Ok(resp) => {
            pool.put(conn, true).await;
            Ok(resp)
        }
        Err(e) => {
            pool.put(conn, false).await;
            Err(e)
        }
    }
}

#[derive(Clone)]
pub enum ForwarderKind {
    Dial,
    Pool(Arc<Manager>),
}

impl ForwarderKind {
    pub async fn forward(&self, addr: &str, req: &Request) -> Result<Response, UpstreamError> {
        match self {
            Self::Dial => super::dial::forward_http(addr, req).await,
            Self::Pool(manager) => {
                let pool = manager
                    .pool(addr)
                    .ok_or_else(|| UpstreamError::Parse(format!("no pool for {addr}")))?;
                forward_pooled(&pool, req).await
            }
        }
    }
}
